### DISCUSSION TITLE
Hierarchical model

### DISCUSSION THREAD

**Discussion by katiacracco** [infer-actively/pymdp#107] [Q&A]
Hello,

has anyone used the pymdp library to build a hierarchical inference model?
I am trying to implement the hierarchical model described in the attached paper, but I don't understand how to use some features of pymdp.

Let me know, thank you!

[Friston_Deeply Felt Affect_ The Emergence of Valence in Deep Active Inference.pdf](https://github.com/infer-actively/pymdp/files/10416954/Friston_Deeply.Felt.Affect_.The.Emergence.of.Valence.in.Deep.Active.Inference.pdf)

**Comment by conorheins** [comment]
Hi @katiacracco,

To my knowledge, no one has published any work using `pymdp` to implement a hierarchical model. As of now, this is a relatively non-standard/ creative use case for `pymdp`, since we do not have any examples of hierarchical POMDPs in the documentation or demos.

However, I recently put together a hierarchical model for a tutorial -- it's not very well-documented right now and a bit ad-hoc, but feel free to use / adapt the code in this notebook as you see fit:

https://colab.research.google.com/drive/1PHVe4cr781YJyVDq4imozx-wbrleU1SC?usp=sharing

It's a recapitulation of the two-level visual foraging model explored in this paper:
[Deep Scene Construction and Active Inference (2020)](https://www.frontiersin.org/articles/10.3389/frai.2020.509354/full)

Hope this helps!

**Reply by katiacracco** [reply]
Hi,

I saw your code and the structure confirms what I deduced from the paper I am working on.
Thank you very much for the comparison!

Have a good day!
